typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1501 @@
+          | None, None -> assert false
+          | Some locs0, Some locs1 -> Some (locs0 @ locs1)
+          (* cannot bind variable only in one of the OR-patterns *)
+          | _, _ -> assert false)
+        ienv0 ienv1
+
+    let empty = Ident.Map.empty
+
+    let conjunct ienv0 ienv1 =
+      Ident.Map.merge
+        (fun _id locs0 locs1 ->
+          match (locs0, locs1) with
+          | None, None -> assert false
+          | Some t, None -> Some t
+          | None, Some t -> Some t
+          (* cannot bind variable twice in a single pattern *)
+          | Some _, Some _ -> assert false)
+        ienv0 ienv1
+
+    let conjuncts = List.fold_left conjunct empty
+    let singleton id locs = Ident.Map.singleton id locs
+  end
+
+  type t = UF.Path.t list Ident.Map.t
+
+  let empty = Ident.Map.empty
+
+  let extend t ex =
+    Ident.Map.merge
+      (fun _id paths0 paths1 ->
+        match (paths0, paths1) with
+        | None, None -> assert false
+        | _, Some paths1 -> Some paths1
+        | Some paths0, None -> Some paths0)
+      t ex
+
+  let find_opt = Ident.Map.find_opt
+end
+
+(** The fun algebraic stuff ends. Here comes the concrete mess *)
+
+(* Forcing due to boundary is more about OCaml than the algebra, hence defined
+   here (as opposed to earlier) *)
+
+(** See [report_error] for explanation *)
+type boundary_reason =
+  | Value_from_mod_class (* currently will never trigger *)
+  | Free_var_of_mod_class (* currently will never trigger *)
+  | Out_of_mod_class
+
+type boundary = {
+  cannot_force : Maybe_unique.cannot_force;
+  reason : boundary_reason;  (** which kind of boundary is being crossed? *)
+}
+
+exception Boundary of boundary
+
+let force_shared_boundary t ~reason =
+  match Maybe_unique.force_shared t with
+  | Ok () -> ()
+  | Error cannot_force -> raise (Boundary { cannot_force; reason })
+
+type single_value_to_match = UF.Path.t list * Location.t * unique_use option
+(** A triple of:
+   - the value's all possible paths.
+   - the location where it's defined
+   - the modes, if it's actually an alias
+*)
+
+type value_to_match =
+  | Match_tuple of single_value_to_match list
+      (** The value being matched is a tuple; we treat it specially so matching
+  tuples against tuples merely create alias instead of uses *)
+  | Match_single of single_value_to_match
+      (** The value being matched is not a tuple *)
+
+let mark_implicit_borrow_memory_address_paths paths ?(access = Occurrence.Read)
+    loc =
+  let occ = Occurrence.mk ~access loc in
+  let mark_one path =
+    (* borrow the memory address of the parent *)
+    UF.singleton
+      (UF.Path.child path Usage_tree.Projection.Memory_address)
+      (* Currently we just generate a dummy unique_barrier ref that won't be
+         consumed. The distinction between implicit and explicit borrowing is
+         still needed because they are handled differently in closures *)
+      (Maybe_shared (Maybe_shared.singleton (ref None) occ))
```

Comment:
I was thinking that `mark_implicit_borrow_memory` is the current abstraction for that. I'm not sure having a function called `maybe_shared` which sounds like identical to `Maybe_shared` but secretly has `singleton`.

Reply:
In the end I only added the one for `Usage.shared`, because this is the only one that's would be used more than once and thus worth adding a shorthand for.